p2p/peer_pool.py

```diff
@@ -124,47 +124,49 @@ def __init__(self,
-                        self.event_bus.request(
-                            PeerCandidatesRequest(available_peer_slots),
-                            TO_DISCOVERY_BROADCAST_CONFIG,
-                        ),
-                        timeout=REQUEST_PEER_CANDIDATE_TIMEOUT
-                    )
-                except TimeoutError:
-                    self.logger.warning("Discovery did not answer PeerCandidateRequest in time")
-                    continue
+    def setup_peer_backends(self) -> Tuple[BasePeerBackend, ...]:
+        return (
+            DiscoveryPeerBackend(self.event_bus),
+            BootnodesPeerBackend(self.event_bus),
+        )
+
+    async def _add_peers_from_backend(self, backend: BasePeerBackend) -> None:
+        available_slots = self.max_peers - len(self)
 
-                # In some cases (e.g ROPSTEN or private testnets), the discovery table might be
-                # full of bad peers so if we can't connect to any peers we try a random bootstrap
-                # node as well.
-                if not len(self):
-                    try:
-                        bootnodes_response = await self.wait(
-                            self.event_bus.request(
-                                RandomBootnodeRequest(),
-                                TO_DISCOVERY_BROADCAST_CONFIG
-                            ),
-                            timeout=REQUEST_PEER_CANDIDATE_TIMEOUT
-                        )
-                    except TimeoutError:
-                        self.logger.warning(
-                            "Discovery did not answer RandomBootnodeRequest in time"
-                        )
-                        continue
-                    candidates = response.candidates + bootnodes_response.candidates
-                else:
-                    candidates = response.candidates
-
-                self.logger.debug2("Received candidates to connect to (%s)", candidates)
-                await self.connect_to_nodes(from_uris(candidates))
+        try:
+            candidates = await self.wait(
+                backend.get_peer_candidates(
+                    num_requested=available_slots,
+                    num_connected_peers=len(self),
+                ),
+                timeout=REQUEST_PEER_CANDIDATE_TIMEOUT,
+            )
+        except TimeoutError:
+            self.logger.warning("PeerCandidateRequest timed out to backend %s", backend)
+            return
+        else:
+            self.logger.debug2(
+                "Got candidates from backend %s (%s)",
+                backend,
+                candidates,
+            )
+            await self.connect_to_nodes(iter(candidates))
+
+    async def maybe_connect_more_peers(self) -> None:
+        # allowed to operate at a maximum rate of 1 check every 2 seconds
+        rate_limiter = token_bucket(0.5, 5)
+        while True:
+            if self.is_full:
+                await self.sleep(DISOVERY_INTERVAL)
```

Comment:
*nit* from old code

```suggestion
                await self.sleep(DISCOVERY_INTERVAL)
```

(plus the import, etc)